crates/sui-core/src/consensus_adapter.rs

```diff
@@ -206,62 +210,76 @@ impl ConsensusAdapter {
         self.num_inflight_transactions.load(Ordering::Relaxed)
     }
 
-    fn should_submit(
+    fn await_submit_delay(
         committee: &Committee,
         ourselves: &AuthorityName,
         transaction: &ConsensusTransaction,
-    ) -> bool {
+    ) -> impl Future<Output = ()> {
+        tokio::time::sleep(Self::submit_delay(committee, ourselves, transaction))
+    }
+
+    fn submit_delay(
+        committee: &Committee,
+        ourselves: &AuthorityName,
+        transaction: &ConsensusTransaction,
+    ) -> Duration {
         if let ConsensusTransactionKind::UserTransaction(certificate) = &transaction.kind {
-            Self::should_submit_certificate(committee, ourselves, certificate.digest())
+            Self::submit_delay_certificate(committee, ourselves, certificate.digest())
         } else {
-            true
+            Duration::ZERO
         }
     }
 
-    /// Check if this authority should submit the certificate to consensus.
-    fn should_submit_certificate(
+    /// Check when this authority should submit the certificate to consensus.
+    fn submit_delay_certificate(
+        committee: &Committee,
+        ourselves: &AuthorityName,
+        tx_digest: &TransactionDigest,
+    ) -> Duration {
+        let (nominator, denominator) =
+            Self::position_submit_certificate(committee, ourselves, tx_digest);
+        const MAX_DELAY_MS: u64 = 60_000;
+        // There is no overflow as long as total_stake < u64::MAX / MAX_DELAY_MS, which seem to be reasonable expectation
+        let duration_ms = MAX_DELAY_MS * nominator / denominator;
```

Comment:
i don't understand why this has to be scaled by stake - if we shuffle by stake (see below), then it would seem correct to simply wait K * i seconds, where K is some constant, and i is our position in the shuffled list.

Reply:
Interesting approach, need to think how it works. You might have some insight about how it addresses the concrete threat outlined in my comment above.

Reply:
The main difference I see is that by doing an unweighted shuffle, every validator has an equal chance of being the first one to submit the cert. The only thing that is weighted in this code is the delay. With a weighted shuffle, heavily weighted validators will be more likely to be the first to submit.

This also seems to produce some odd behavior. For instance if you have the following permutation of validators,stake tuples: 

(A, 100)
(B, 1)
(C, 1)

A will submit immediately, and then B and C will both wait nearly the entire max delay, whereas it would seem better to have B submit at 1/2 of max delay. Admittedly this is a contrived example and we probably won't see such lopsided weights - I'm just trying to show that the intuition behind scaling delay by stake is odd.

Regarding complexity, https://docs.rs/rand/latest/rand/seq/trait.SliceRandom.html#tymethod.choose_multiple_weighted claims to be O(n) time if the partition_at_index feature is enabled, and that feature appears to have been stabilized years ago.